File: src/main/java/com/donrobo/fpbg/blueprint/building/Building.java
package com.donrobo.fpbg.blueprint.building;

import com.donrobo.fpbg.blueprint.Direction;
import com.donrobo.fpbg.data.Int2;
import net.sf.json.JSONObject;

public interface Building {
    JSONObject toJson();

    int getWidth();

    int getHeight();

    int getX();

    int getY();

    Building move(int x, int y);

    String getName();

    static <T extends Building> T move(T building, int offset, Direction direction) {
        Int2 offsetInt2 = direction.move(new Int2(0, 0), offset);

        Building movedBuilding = building.move(offsetInt2.getX(), offsetInt2.getY());
        return (T) movedBuilding;
    }

    Int2 getPosition();

    char getVisualizationCharacter();
}

File: src/main/kotlin/com/donrobo/fpbg/blueprint/building/Building.kt
package com.donrobo.fpbg.blueprint.building

import com.donrobo.fpbg.blueprint.Direction
import com.donrobo.fpbg.data.Int2
import net.sf.json.JSONObject

interface Building {
    fun toJson(): JSONObject

    val width: Int

    val height: Int

    val x: Int

    val y: Int

    fun move(x: Int, y: Int): Building

    val name: String

    val position: Int2

    val visualizationCharacter: Char

    fun rotateCCW(around: Int2, count: Int): Building

    fun move(offset: Int, direction: Direction): Building
}
